python/plugins/processing/algs/qgis/RasterCalculator.py

```diff
@@ -122,18 +107,19 @@ def processAlgorithm(self, parameters, context, feedback):
 
         layersDict = {}
         if layers:
-            layersDict = {os.path.basename(lyr.source().split(".")[0]): lyr for lyr in layers}
+            layersDict = {lyr.source(): lyr for lyr in layers}
 
         crs = self.parameterAsCrs(parameters, self.CRS, context)
-        if not layers and not crs.isValid():
-            raise QgsProcessingException(self.tr("No reference layer selected nor CRS provided"))
-
-        if not crs.isValid() and layers:
-            crs = list(layersDict.values())[0].crs()
+        if not crs or not crs.isValid():
```

Comment:
I think this may need to be 

    if crs is None or not crs.isValid():

(Because I'm not sure what boolean check sip will apply to QgsCoordinateReferenceSystem, so I think it's safer to be explicit here)